### PR TITLE
fix partial schema with list input

### LINKED ISSUE #10943: Partial schema in from_dicts doesn't work any longer?
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python

In [1]: import polars as pl

In [2]: pl.__version__
Out[2]: '0.19.2'

In [3]: pl.from_dicts([{"a": 1, "b": 2}], schema=["a"])
Out[3]: 
shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
└─────┘

In [4]: pl.from_dicts([{"a": 1, "b": 2}], schema=["b"])
---------------------------------------------------------------------------
ShapeError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 pl.from_dicts([{"a": 1, "b": 2}], schema=["b"])

In [5]: pl.from_dicts([{"a": 1, "b": 2}], schema={"a": None})
Out[5]: 
shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
└─────┘

In [6]: pl.from_dicts([{"a": 1, "b": 2}], schema={"b": None})
---------------------------------------------------------------------------
ShapeError                                Traceback (most recent call last)
Cell In[6], line 1

```

### Log output

_No response_

### Issue description

>If you want to drop some of the fields found in the input dictionaries, a _partial_ schema can be declared, in which case omitted fields will not be loaded. Similarly, you can extend the loaded frame with empty columns by adding them to the schema.

### Expected behavior

Output for partial `b` should be like partial `a`, keeping only the `b` column

### Installed versions

<details>

```
In [3]: pl.show_versions()
--------Version info---------
Polars:              0.19.2
Index type:          UInt32
Platform:            macOS-13.4-arm64-arm-64bit
Python:              3.11.3 (main, May  7 2023, 17:43:38) [Clang 16.0.3 ]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.9.0
matplotlib:          3.7.2
numpy:               1.25.1
pandas:              2.0.3
pyarrow:             13.0.0
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>

```

</details>

### PR DESCRIPTION
**Pull request by hxzhao527**
fix #10943

### CODE REVIEW

File: py-polars/polars/utils/_construction.py [hunk 1]
```diff
@@ -622,11 +622,12 @@ def _post_apply_columns(
         (columns or pydf_columns), schema_overrides=schema_overrides
     )
     column_subset: list[str] = []
-    if columns != pydf_columns:
-        if len(columns) < len(pydf_columns) and columns == pydf_columns[: len(columns)]:
-            column_subset = columns
-        else:
-            pydf.set_column_names(columns)
+    if set(pydf_columns).issuperset(columns):
```
Comment (ritchie46) [inline py-polars/polars/utils/_construction.py:625]: Could we come up with a cheaper check? One that doesn't require us to create a full set.

Comment (hxzhao527) [reply py-polars/polars/utils/_construction.py:625]: Before this check, we just call `_unpack_schema` which may return a list no order fixed.

https://github.com/pola-rs/polars/blob/d3721f15167c080df6c843bb6d8b07105e1fc932/py-polars/polars/utils/_construction.py#L613-L623

https://github.com/pola-rs/polars/blob/d3721f15167c080df6c843bb6d8b07105e1fc932/py-polars/polars/utils/_construction.py#L651-L665

So the problem becomes that to check two series if one is supperset of the other which  one order not fixed.

We can sort both, then check if one is subslice of the other, the time complexity will be $O(nlogn + mlogm + min(m, n))$ at least(sort and compare), no more or a const  extra space required.

`set(pydf_columns).issuperset(columns)` or `set(columns).issubset(pydf_columns)` , the time complexity will be $O(m+ n)$ (convert and compare), $O(m)$ or $O(n)$ extra space required.

**I prefer _the less time_ way.**

we also can do some pre-check like this.
```python
    if len(columns) < len(pydf_columns) and set(pydf_columns).issuperset(columns):
        column_subset = columns
    elif len(columns) == len(pydf_columns):
        if set(pydf_columns).issuperset(columns):
            column_subset = columns
        else:
            pydf.set_column_names(columns)
    else:
        raise ShapeError(
            f"{len(columns)} column names provided "
            f"for a dataframe of width {len(pydf_columns)}"
        )
```

File: py-polars/tests/unit/datatypes/test_decimal.py [hunk 2]
```diff
@@ -16,7 +16,7 @@ def permutations_int_dec_none() -> list[tuple[D | int | None, ...]]:
                 D("-0.01"),
                 D("1.2345678"),
                 D("500"),
-                -1,
+                D(-1),
```
Comment (hxzhao527) [inline py-polars/tests/unit/datatypes/test_decimal.py:19]: irrelevant to  partial schema. but test [`test_series_from_pydecimal_and_ints` ](https://github.com/pola-rs/polars/blob/d3721f15167c080df6c843bb6d8b07105e1fc932/py-polars/tests/unit/datatypes/test_decimal.py#L26) failed.

File: py-polars/polars/utils/_construction.py (hunk 1 above)
Comment (ritchie46) [reply py-polars/polars/utils/_construction.py:625]: I definitely want the pre-checks.

This is not by doing of you, but I didn't realize we can be so expensive after creating the `DataFrame`. I think we should look if we can hoist some work on to the Rust side, or elide that here. :thinking:

**Comment by ritchie46** [comment]
@alexander-beedie could you take a look here? Deep into your territory ;)

**Comment by alexander-beedie** [comment]
> @alexander-beedie could you take a look here? Deep into your territory ;)

No problem :)  Can get to it this afternoon (I see there's a proposed optimisation in a similar area too, so will look at both) 👍

File: py-polars/polars/utils/_construction.py (hunk 1 above)
Comment (hxzhao527) [reply py-polars/polars/utils/_construction.py:625]: I get it. I will see how to make it happen in Rust.

**Comment by alexander-beedie** [comment]
Ok, I've seen a _very_ cheap way (a single `isinstance` check) to have the Rust-side `read_dicts` constructor do the right thing here, with no need to engage Python-side column checking at all.

The only downside is that we will technically have to break a weird override that we currently support - however, I would be strongly in favour of breaking it as it looks much more likely to cause accidental errors than be genuinely useful. 

Specifically:
```python
dicts = [
    {"a": 1, "b": 2},
    {"b": 1, "a": 2},
    {"a": 1, "b": 2},
]
pl.DataFrame(dicts, schema=["c", "d"])
# shape: (3, 2)
# ┌─────┬─────┐
# │ c   ┆ d   │   # !!!
# │ --- ┆ --- │
# │ i64 ┆ i64 │
# ╞═════╪═════╡
# │ 1   ┆ 2   │
# │ 2   ┆ 1   │
# │ 1   ┆ 2   │
# └─────┴─────┘
```
The number of ways this type of override could go wrong is... non-trivial.
It would be much better expressed as an _explicit_ rename:
```python
pl.DataFrame(dicts).rename({"a": "c", "b": "d"})
```

@ritchie46: Thoughts? 

Pros: It will make dict init more reliable and completely avoid additional Python-side column-checking in the case of partial schemas.
Cons: We break a somewhat sketchy-looking override 🤔

File: py-polars/polars/utils/_construction.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/utils/_construction.py:625]: > I get it. I will see how to make it happen in Rust.

Nice - but don't do anything yet; I think we can avoid doing anything in Python _without_ requiring any new Rust (for this case), if we accept losing one unusual/awkward override that we technically support at the moment ;)

**Comment by alexander-beedie** [comment]
> In other word, if data type is `list[dict]`, argument schema will be ignored?

No, in practice it's actually quite useful to be able to partially-load dicts.

**Comment by ritchie46** [comment]
@alexander-beedie I am totally in favor of that. I think we can just break as the previous didn't work and I wonder allowed if we intended that behavior. 🤔

**Comment by alexander-beedie** [comment]
> @alexander-beedie I am totally in favor of that. I think we can just break as the previous didn't work and I wonder allowed if we intended that behavior. 🤔 

We may have inadvertently dragged this behaviour along with us; since we can achieve the same result much more sensibly by renaming (or just setting `df.columns` directly) I think treating it as a bug fix makes sense.

Will fix tomorrow evening; having a laptop-free weekend travelling up to Dubai 😅

**Comment by alexander-beedie** [comment]
@hxzhao527 I'll make a separate small PR tomorrow to fix this a little differently, but then I think you can rebase and we should be able to use this PR to apply some of your other proposed optimisations. Thanks for looking at it! 👍

**Comment by alexander-beedie** [comment]
Ok, the fix that defers to Rust for (partial) schema determination is ready (ref: https://github.com/pola-rs/polars/pull/11047). 

@hxzhao527: I'll close this PR as being superseded by that one, but if you would like to open a new PR (once that one merges) focusing on improving some of the performance and reducing initialisation latency that would be very welcome! 👍

**Comment by alexander-beedie** [comment]
(Superseded by https://github.com/pola-rs/polars/pull/11047).

**Pull request closed without merging**